src/core/lib/event_engine/windows/windows_listener.cc

```diff
@@ -0,0 +1,364 @@
+  // will still get an IOCP notification for it. So let's just ignore it.
+  if (success != 0) {
+    int last_error = WSAGetLastError();
+    if (last_error != ERROR_IO_PENDING) {
+      return fail(GRPC_WSA_ERROR(last_error, "AcceptEx"));
+    }
+  }
+  // We're ready to do the accept. Calling NotifyOnRead may immediately process
+  // an accept that happened in the meantime.
+  io_state_->accept_socket = accept_socket;
+  io_state_->listener_socket->NotifyOnRead(&io_state_->on_accept_cb);
+  GRPC_EVENT_ENGINE_TRACE(
+      "SinglePortSocketListener::%p listening. listener_socket::%p", this,
+      io_state_->listener_socket.get());
+  return absl::OkStatus();
+}
+
+void WindowsEventEngineListener::SinglePortSocketListener::
+    OnAcceptCallbackLocked() {
+  auto close_socket_and_restart =
+      [&](bool do_close_socket = true)
+          ABSL_EXCLUSIVE_LOCKS_REQUIRED(io_state_->mu) {
+            if (do_close_socket) closesocket(io_state_->accept_socket);
+            GPR_ASSERT(GRPC_LOG_IF_ERROR("SinglePortSocketListener::Start",
+                                         StartLocked()));
+          };
+  const auto& overlapped_result =
+      io_state_->listener_socket->read_info()->result();
+  if (overlapped_result.wsa_error != 0) {
+    gpr_log(GPR_ERROR, "%s",
+            GRPC_WSA_ERROR(overlapped_result.wsa_error,
+                           "Skipping on_accept due to error")
+                .ToString()
+                .c_str());
+    return close_socket_and_restart();
+  }
+  SOCKET tmp_listener_socket = io_state_->listener_socket->raw_socket();
+  int err =
+      setsockopt(io_state_->accept_socket, SOL_SOCKET, SO_UPDATE_ACCEPT_CONTEXT,
+                 reinterpret_cast<char*>(&tmp_listener_socket),
+                 sizeof(tmp_listener_socket));
+  if (err != 0) {
+    gpr_log(GPR_ERROR, "%s",
+            GRPC_WSA_ERROR(WSAGetLastError(), "setsockopt").ToString().c_str());
+    return close_socket_and_restart();
+  }
+  EventEngine::ResolvedAddress peer_address;
+  int peer_name_len = EventEngine::ResolvedAddress::MAX_SIZE_BYTES;
+  err = getpeername(io_state_->accept_socket,
+                    const_cast<sockaddr*>(peer_address.address()),
+                    &peer_name_len);
+  if (err != 0) {
+    gpr_log(
+        GPR_ERROR, "%s",
+        GRPC_WSA_ERROR(WSAGetLastError(), "getpeername").ToString().c_str());
+    return close_socket_and_restart();
+  }
+  peer_address =
+      EventEngine::ResolvedAddress(peer_address.address(), peer_name_len);
+  auto addr_uri = ResolvedAddressToURI(peer_address);
+  std::string peer_name = "unknown";
+  if (!addr_uri.ok()) {
+    // TODO(hork): test an early exit/restart here with end2end tests
+    gpr_log(GPR_ERROR, "invalid peer name: %s",
+            addr_uri.status().ToString().c_str());
+  } else {
+    peer_name = *addr_uri;
+  }
+  auto endpoint = std::make_unique<WindowsEndpoint>(
+      peer_address, listener_->iocp_->Watch(io_state_->accept_socket),
```

Comment:
Would it be prudent to do `io_state_->accept_socket = INVALID_SOCKET;` (maybe in `close_socket_and_restart`) after we hand over this socket to the endpoint? I see because of the `GPR_ASSERT` on line 153 the next `StartLocked()` call must succeed and `io_state_->accept_socket` would be overridden, so there is no chance we would close a socket that has already been given to an endpoint. Just wondering...

Reply:
Sure, done. I'm not sure the race is possible, maybe an in-use `accept_socket` could get closed inadvertently if the listener and its SPLs are shut down between handing the accept_socket off to the endpoint, before StartLocked is called. Even if that's not possible now, I'm fine with the added caution and future-proofing.